File: Rpkg/semx/tests/testthat/test-frontend-integration.R
test_that("semx_model round-trips Laplace GLMM", {
  model <- semx_model(
    equations = c("y ~ 1 + cluster"),
    families = c(y = "binomial"),
    kinds = c(cluster = "grouping"),
    covariances = list(
      list(name = "G_diag", structure = "diagonal", dimension = 1)
    ),
    random_effects = list(
      list(name = "u_cluster", variables = c("cluster"), covariance = "G_diag")
    )
  )
  
  ir <- model$ir
  expect_equal(length(ir$covariances), 1)
  expect_equal(ir$covariances[[1]]$id, "G_diag")
  expect_equal(ir$covariances[[1]]$structure, "diagonal")
  
  expect_equal(length(ir$random_effects), 1)
  expect_equal(ir$random_effects[[1]]$id, "u_cluster")
  expect_equal(ir$random_effects[[1]]$variables, c("cluster"))
  expect_equal(ir$random_effects[[1]]$covariance_id, "G_diag")
})

test_that("semx_model round-trips Survival", {
  model <- semx_model(
    equations = c("y ~ x"),
    families = c(y = "weibull", x = "gaussian")
  )
  
  ir <- model$ir
  vars <- ir$variables
  
  # Helper to find variable by name
  find_var <- function(name) {
    for (v in vars) {
      if (v$name == name) return(v)
    }
    NULL
  }
  
  y_var <- find_var("y")
  expect_false(is.null(y_var))
  expect_equal(y_var$family, "weibull")
  
  x_var <- find_var("x")
  expect_false(is.null(x_var))
  # VariableKind.Observed is 0
  expect_equal(x_var$kind, 0) 
})

test_that("semx_model round-trips Kronecker", {
  model <- semx_model(
    equations = c("y ~ 1 + env + geno"),
    families = c(y = "gaussian"),
    kinds = c(env = "grouping", geno = "grouping"),
    covariances = list(
      list(name = "G_kron", structure = "kronecker", dimension = 1)
    ),
    random_effects = list(
      list(name = "u_gxe", variables = c("env", "geno"), covariance = "G_kron")
    )
  )
  
  ir <- model$ir
  expect_equal(length(ir$covariances), 1)
  expect_equal(ir$covariances[[1]]$id, "G_kron")
  expect_equal(ir$covariances[[1]]$structure, "kronecker")
  
  expect_equal(length(ir$random_effects), 1)
  expect_equal(ir$random_effects[[1]]$id, "u_gxe")
  expect_equal(ir$random_effects[[1]]$variables, c("env", "geno"))
})

File: python/semx/tests/test_frontend_integration.py
"""Integration tests for the semx.Model front-end."""

import pytest
from semx import Model, VariableKind

def test_frontend_roundtrip_laplace_glmm():
    """
    Round-trip a Binomial GLMM (Laplace) model.
    Corresponds to: y ~ 1 + (1 | cluster)
    """
    model = Model(
        equations=["y ~ 1 + cluster"],
        families={"y": "binomial"},
        kinds={"cluster": "grouping"},
        covariances=[
            {"name": "G_diag", "structure": "diagonal", "dimension": 1}
        ],
        random_effects=[
            {"name": "u_cluster", "variables": ["cluster"], "covariance": "G_diag"}
        ]
    )
    
    ir = model.to_ir()
    
    # Check variables
    var_map = {v.name: v for v in ir.variables}
    assert var_map["y"].family == "binomial"
    assert var_map["cluster"].kind == VariableKind.Grouping
    
    # Check covariances
    assert len(ir.covariances) == 1
    assert ir.covariances[0].id == "G_diag"
    assert ir.covariances[0].structure == "diagonal"
    
    # Check random effects
    assert len(ir.random_effects) == 1
    assert ir.random_effects[0].id == "u_cluster"
    assert ir.random_effects[0].variables == ["cluster"]
    assert ir.random_effects[0].covariance_id == "G_diag"

def test_frontend_roundtrip_survival():
    """
    Round-trip a Survival model (Weibull).
    Corresponds to: y ~ x
    """
    model = Model(
        equations=["y ~ x"],
        families={"y": "weibull", "x": "gaussian"},
    )
    ir = model.to_ir()
    
    # Check variables
    var_map = {v.name: v for v in ir.variables}
    assert var_map["y"].family == "weibull"
    assert var_map["x"].kind == VariableKind.Observed

def test_frontend_roundtrip_kronecker():
    """
    Round-trip a Kronecker covariance model.
    """
    model = Model(
        equations=["y ~ 1 + env + geno"],
        families={"y": "gaussian"},
        kinds={"env": "grouping", "geno": "grouping"},
        covariances=[
            {"name": "G_kron", "structure": "kronecker", "dimension": 1}
        ],
        random_effects=[
            {"name": "u_gxe", "variables": ["env", "geno"], "covariance": "G_kron"}
        ]
    )
    
    ir = model.to_ir()
    
    assert len(ir.covariances) == 1
    assert ir.covariances[0].id == "G_kron"
    assert ir.covariances[0].structure == "kronecker"
    
    assert len(ir.random_effects) == 1
    assert ir.random_effects[0].id == "u_gxe"
    assert ir.random_effects[0].variables == ["env", "geno"]
    assert ir.random_effects[0].covariance_id == "G_kron"
